### PR TITLE
feat: Program Screen Error Handling

### PR DESCRIPTION
**Pull request by shafqat-muneer**
[LEARNER-9992](https://2u-internal.atlassian.net/browse/LEARNER-9992): feat: Program Screen Error Handling

| Light Mode | Dark Mode |
|----------|----------|
| Simulator Screenshot - iPhone 15 - 2024-06-03 at 10 53 51 | Simulator Screenshot - iPhone 15 - 2024-06-03 at 10 53 48 |

### CODE REVIEW

File: Core/Core/View/Base/Webview/WebView.swift [hunk 1]
```diff
@@ -17,6 +17,8 @@ public protocol WebViewNavigationDelegate: AnyObject {
         shouldLoad request: URLRequest,
         navigationAction: WKNavigationAction
     ) async -> Bool
+    
+    func showWebViewError()
```
Comment (saeedbashir) [inline Core/Core/View/Base/Webview/WebView.swift:21]: It nice and can work with passing the error, but how about having a more detailed method like 
```
func webView(
        _ webView: WKWebView,
        didFail navigation: WKNavigation,
        withError error: Error
    )
 ```   
Thoughts?

File: Core/Core/View/Base/Webview/WebView.swift [hunk 2]
```diff
@@ -189,7 +199,14 @@ public struct WebView: UIViewRepresentable {
         
         @objc private func reload() {
             parent.isLoading = true
-            webview?.reload()
+            if webview?.url?.absoluteString.isEmpty ?? true {
+                if let url = URL(string: parent.viewModel.url) {
```
Comment (saeedbashir) [inline Core/Core/View/Base/Webview/WebView.swift:203]: What will happen if this condition fails? will it reload the webview?

File: Course/Course.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -526,13 +525,20 @@
 			path = Mock;
 			sourceTree = "<group>";
 		};
-		97CA95212B875EA200A9EDEA /* Views */ = {
+		9784D4752BF39EEF00AFEFFF /* CalendarSyncProgressView */ = {
```
Comment (saeedbashir) [inline Course/Course.xcodeproj/project.pbxproj:528]: Any reason to update this? Asking because it doesn't seem relevant to the work.

**Review by saeedbashir — CHANGES_REQUESTED** [review]
* Discovery is not showing the full screen erros
* It would be nice if it central align. and any reason to hide reload button here?
image

File: Core/Core/View/Base/Webview/WebView.swift (hunk 1 above)
Comment (shafqat-muneer) [reply Core/Core/View/Base/Webview/WebView.swift:21]: In our use case, displaying specific error details is not required. We are currently only showing internet error or generic error. Therefore, detailed information is not required at this time. We can reconsider this approach if there arises a need for more detailed error messages in the future.

File: Core/Core/View/Base/Webview/WebView.swift (hunk 2 above)
Comment (shafqat-muneer) [reply Core/Core/View/Base/Webview/WebView.swift]: Thank you for bringing this in attention. It’s fixed.

File: Course/Course.xcodeproj/project.pbxproj (hunk 3 above)
Comment (shafqat-muneer) [reply Course/Course.xcodeproj/project.pbxproj:528]: **Updated location from:**
```
Course/Views/CalendarSyncProgressView.swift
Course/Views/DatesSuccessView.swift
```
**To:**
```
Course/Presentation/Subviews/CalendarSyncProgressView/CalendarSyncProgressView.swift
Course/Presentation/Subviews/DatesSuccessView/DatesSuccessView.swift
```

I found that we already have subviews folder and it seems more appropriate to place it there.

**Comment by shafqat-muneer** [comment]
@saeedbashir Feedback addressed. Ready for another round of review 🎉

File: Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                Spacer()
+                switch errorType {
+                case .noInternet, .noInternetWithReload:
+                    CoreAssets.noWifi.swiftUIImage
+                        .renderingMode(.template)
+                        .foregroundStyle(Color.primary)
+                        .scaledToFit()
+                    
+                    Text(CoreLocalization.Error.Internet.noInternetTitle)
+                                    .font(Theme.Fonts.titleLarge)
```
Comment (rnr) [inline Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift:48]: too big indent

File: Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                        .multilineTextAlignment(.center)
+                        .padding(.horizontal, 50)
+                case .generic:
+                    CoreAssets.notAvaliable.swiftUIImage
+                        .renderingMode(.template)
+                        .foregroundStyle(Color.primary)
+                        .scaledToFit()
+                    
+                    Text(CoreLocalization.View.Snackbar.tryAgainBtn)
+                                    .font(Theme.Fonts.titleLarge)
```
Comment (rnr) [inline Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift:63]: fix indent here please

File: Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            .frame(maxWidth: .infinity, maxHeight: proxy.size.height)
+            .background(
+                Theme.Colors.background
+            )
+        }
+    }
+}
+
+#if DEBUG
+//swiftlint:disable all
```
Comment (rnr) [inline Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift:89]: which swiftlint's rules we disable here?

File: Discovery/Discovery/Presentation/WebPrograms/ProgramWebviewView.swift [hunk 7]
```diff
@@ -114,3 +130,25 @@ public struct ProgramWebviewView: View {
         .animation(.default, value: viewModel.showError)
     }
 }
+
+#if DEBUG
+//swiftlint:disable all
```
Comment (rnr) [inline Discovery/Discovery/Presentation/WebPrograms/ProgramWebviewView.swift:135]: why do we need to disable swiftlint here?

**Comment by rnr** [comment]
@moiz994  do we need this yellow line on Programs and Discover screens?

**Comment by rnr** [comment]
@shafqat-muneer 
When the connection is restored and we get the data on the Program screen (for example) and then go to the Discover screen, there we see a different error ("something went wrong" instead of "no connection"). Is this expected?


https://github.com/openedx/openedx-app-ios/assets/37253/d848e18a-5103-4af2-8c64-7724f40c9d65

File: Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift (hunk 6 above)
Comment (shafqat-muneer) [reply Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift:89]: Thanks for bringing that to my attention. I've removed it.

File: Discovery/Discovery/Presentation/WebPrograms/ProgramWebviewView.swift (hunk 7 above)
Comment (shafqat-muneer) [reply Discovery/Discovery/Presentation/WebPrograms/ProgramWebviewView.swift:135]: Thanks for bringing that to my attention. I've removed it.

**Comment by shafqat-muneer** [comment]
> @moiz994 do we need this yellow line on Programs and Discover screens?
> 
> 

It appears to be redundant after implementation of full screen error. Just removed it from Programs and Discovery screens.

**Comment by shafqat-muneer** [comment]
> @shafqat-muneer When the connection is restored and we get the data on the Program screen (for example) and then go to the Discover screen, there we see a different error ("something went wrong" instead of "no connection"). Is this expected?
> 
>  RPReplay_Final1717765154.MOV

In this scenario, the connection has already been restored, so we shouldn't display the "no connection" alert. Instead, it's showing a generic error message. Currently, the user has to tap the reload button to refresh the screen. If we can enhance the user experience, we should consider exploring alternative approaches.

**Comment by rnr** [comment]
@shafqat-muneer After the latest changes, I see the following:
- start the application with a connection
- disconnect the connection
- go to Discover: something is displayed (possibly cached data), but courses are not available
- go to Programmes: Infinite download
So it looks like the "No internet connection" view is now not displayed at all

https://github.com/openedx/openedx-app-ios/assets/37253/63d577ab-7e9c-4d19-88e2-4764a657cea6

Please test this on your side
Thank you

**Comment by shafqat-muneer** [comment]
> @shafqat-muneer When the connection is restored and we get the data on the Program screen (for example) and then go to the Discover screen, there we see a different error ("something went wrong" instead of "no connection"). Is this expected?
> 
>  RPReplay_Final1717765154.MOV

@rnr It's fixed.

**Comment by shafqat-muneer** [comment]
@rnr @saeedbashir PR is prepared for another round of review. 🎉

**Review by saeedbashir — DISMISSED** [review]
LGTM 👍

**Review by rnr — APPROVED** [review]

**Review by rnr — DISMISSED** [review]

File: Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    
+    public init(
+        type: ErrorType
+    ) {
+        self.errorType = type
+    }
+    
+    public init(
+        type: ErrorType,
+        reloadAction: @escaping () -> Void
```
Comment (volodymyr-chekyrta) [inline Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift:30]: Could we generalize it for the future and call it just `action`?

File: Core/Core/View/Base/Webview/WebView.swift [hunk 9]
```diff
@@ -78,6 +87,10 @@ public struct WebView: UIViewRepresentable {
             withError error: Error
         ) {
             webView.isHidden = false
+            DispatchQueue.main.async {
+                self.parent.isLoading = false
+            }
+            parent.webViewNavDelegate?.showWebViewError()
```
Comment (volodymyr-chekyrta) [inline Core/Core/View/Base/Webview/WebView.swift:93]: what about
```
            DispatchQueue.main.async {
                self.parent.isLoading = false
                self.parent.webViewNavDelegate?.showWebViewError()
            }
```
and remove DispatchQueue.main from the showWebViewError function.
wdyt?

File: Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift (hunk 8 above)
Comment (shafqat-muneer) [reply Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift]: Yes, it's updated with a more generic name.

File: Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                        .font(Theme.Fonts.bodyLarge)
+                        .foregroundColor(Theme.Colors.textPrimary)
+                        .multilineTextAlignment(.center)
+                        .padding(.horizontal, 50)
+                }
+                
+                if errorType != .noInternet {
+                    UnitButtonView(type: .reload, action: {
+                        self.reloadAction()
+                    })
```
[the remark this answers is not in the document]
Comment (shafqat-muneer) [reply Core/Core/View/Base/FullScreenErrorView/FullScreenErrorView.swift]: Done 🎉

File: Core/Core/View/Base/Webview/WebView.swift (hunk 9 above)
Comment (shafqat-muneer) [reply Core/Core/View/Base/Webview/WebView.swift]: Agreed, it's more appropriate I believe. Thanks

**Comment by shafqat-muneer** [comment]
@volodymyr-chekyrta Incorporated the suggestions. Ready for another review. 🎉

**Review by volodymyr-chekyrta — APPROVED** [review]
LGTM 👍
